### DISCUSSION TITLE
Use browser-storage to support online/offline editing

### DISCUSSION THREAD

**Discussion by fkmiec** [TiddlyWiki/TiddlyWiki5#7167] [Ideas]
Hi all. 

I have been using TiddlyWiki for about a year now and am a big fan. Amazing work. 

Like many users, I've tried quite a few syncer and saver options. Ultimately, I prefer the efficiency of the syncer mechanism used with Node JS and the TiddlyWeb implementations. One thing I've found is that it is possible to lose connection with the server (bad wifi connection, server down for maintenance, start at home where the server is and going mobile for the day) and potentially to lose work under that circumstance. I realize that the sync mechanism will sync everything when reconnected as long as my browser remains open, but if it closes or I accidentally navigate away, my changes can be lost.  

The browser-storage plugin provides the basis for saving locally when the server is not available, but if the browser is closed or you navigate away, the tiddlers saved in browser storage are not lost, but they do not sync to the server when connectivity is restored. 

I created a [browser-storage-sync plugin](https://github.com/fkmiec/tiddlywiki-plugin-tiddlywebadapter-browser-storage-sync) that handles re-syncing the tiddlers from browser storage to the server and removing a tiddler from browser storage whenever sync is complete. It is also conditional on the presence of the browser-storage plugin and is otherwise inert. These were two relatively small changes which I added to the tiddlywebadaptor.js file. I don't know if there was an alternative way to implement this functionality, but it seemed like the most appropriate place to put it. The downside, of course, is that I think tiddlywebadaptor is pretty frequently modified, so implementing as an additional plugin might run into breaking changes or conflicts with yet other plugins. 

I'm curious if those of you developing the core and the tiddlywebadaptor plugin would consider including this functionality directly in the tiddlywebadaptor? It strikes me as something that would benefit anyone using the server mode. 

On a related note, I created an [alert-suppression plugin](https://github.com/fkmiec/tiddlywiki-plugin-alertsuppression) that suppresses the alert banner if it repeats, as it does when the server connection is broken. It shows the banner initially to alert the user and the sync icon continues to show red, but the banner then gets out of the way so you can edit without the banner interfering. I think this, too, is a good idea that would be great to address in the core. Thoughts?

**Comment by Jermolene** [comment]
Hi @fkmiec welcome to the project, and thank you for the kind words.

Integrating the browser-storage plugin with the TiddlyWeb sync adaptor is an excellent idea, and I do think it would make sense for the core. I like the idea of the two plugins working together.

I notice that your code overwrites tiddlywebadaptor.js. It would be easier to see the modifications if you could mark them, or perhaps make a draft PR with the changes so that we can review them.

**Reply by fkmiec** [reply]
Thanks @Jermolene. I submitted a pull request.
